recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1150 @@
+    ) -> Optimizer:
+        optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
+        if opt_state_dict:
+            training.load_from_full_optimizer_state_dict(
+                optimizer,
+                opt_state_dict,
+                self._device,
+            )
+
+        if self._is_rank_zero:
+            log.info("Optimizer is initialized.")
+        return optimizer
+
+    def _setup_lr_scheduler(
+        self,
+        cfg_lr_scheduler: DictConfig,
+        num_training_steps: int,
+        last_epoch: int,
+    ) -> Optimizer:
+        lr_scheduler = config.instantiate(
+            cfg_lr_scheduler,
+            self._optimizer,
+            num_training_steps=num_training_steps,
+            last_epoch=last_epoch,
+        )
+        if self._is_rank_zero:
+            log.info("Learning rate scheduler is initialized.")
+        return lr_scheduler
+
+    def _setup_one_dataset(
+        self,
+        cfg_dataset: DictConfig,
+        global_shuffle: bool,
+        global_parallel_method: str,
+        global_streaming: bool,
+        global_num_workers: int,
+    ) -> BaseNode:
+        streaming = cfg_dataset.pop("streaming", global_streaming)
+        parallel_method = cfg_dataset.pop("parallel_method", global_parallel_method)
+        shuffle = cfg_dataset.pop("shuffle", global_shuffle)
+        num_workers = cfg_dataset.pop("num_workers", global_num_workers)
+
+        return config.instantiate(
+            cfg_dataset,
+            self._tokenizer,
+            streaming=streaming,
+            shuffle=shuffle,
+            parallel_method=parallel_method,
+            num_workers=num_workers,
+        )
+
+        return ds
+
+    def _setup_data_td(
+        self,
+        cfg_dataloader: DictConfig,
+        cfg_multi_datasets: DictConfig,
+        batch_size: int,
+    ) -> Loader:
+        """
+        All torchdata related setup happens here. Currently this recipe supports
+        both Map and Streaming datasets (from HuggingFace datasets), and mixing multiple
+        datasets (can be mix of Map and Streaming).
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        # Get global settings
+        shuffle = cfg_dataloader.shuffle
+        parallel_method = cfg_dataloader.get("parallel_method", "thread")
+        packed = cfg_dataloader.get("packed", False)
+        streaming = cfg_dataloader.get("streaming", False)
+        num_workers = cfg_dataloader.get("num_workers", 0)
+        pin_memory = cfg_dataloader.get("pin_memory", True)
+        collate_fn = cfg_dataloader.collate_fn
+        prefetch_factor = cfg_dataloader.get("prefetch_factor", 2)
+
+        stop_criterion = cfg_multi_datasets.get(
+            "stop_criterion", StopCriteria.CYCLE_UNTIL_ALL_DATASETS_EXHAUSTED
+        )
+        weights, datasets = {}, {}
+        cfg_datasets = cfg_multi_datasets.datasets
+        for k, cfg_and_weight in cfg_datasets.items():
```

Comment:
all of this logic needs to be in the builder. we do not want to expose torchdata internals in each recipe. I am totally okay with creating a new file in `torchtune/data` that contains functions that set up torchdata dataloaders and nodes.

Another ideal that I'm curious if we can achieve, can we unify the UX for multi datasets and single datasets? i.e., if we had a `get_dataloader` method, you can pass in a single dataset or a multi dataset and the call is the same in the recipe regardless of what the user specifies in the config